AccommodationService/Infrastructure/Services/AvailabilityPeriodService.cs

```diff
@@ -0,0 +1,52 @@
+﻿using AccommodationService.Application.Repositories;
+using AccommodationService.Application.Services;
+using AccommodationService.Domain;
+
+namespace AccommodationService.Infrastructure.Services;
+
+public class AvailabilityPeriodService : IAvailabilityPeriodService
+{
+    private readonly IAvailabilityPeriodRepository availabilityPeriodRepository;
+
+    public AvailabilityPeriodService(IAvailabilityPeriodRepository availabilityPeriodRepository)
+    {
+        this.availabilityPeriodRepository = availabilityPeriodRepository;
+    }
+
+    public async Task<AvailabilityPeriod> CreateAsync(AvailabilityPeriod availabilityPeriod)
+    {
+        var closestAvailabilityPeriod = await availabilityPeriodRepository
+            .FindClosestStartDateAsync(availabilityPeriod.PropertyId, availabilityPeriod.StartDate);
+
+        if (closestAvailabilityPeriod != null && closestAvailabilityPeriod.EndDate >= availabilityPeriod.StartDate)

```

Comment:
I guess we should enrich this check?

Finding closes start date returns dates that are starting before the date of availability period being created.
But what happens with periods that are starting after the start date of the creating period but they also start before end of creating period.

Example:
You have period from 10.05.-20.05.
You try to create period from 12.05.-X.X. your check will work. It will return the period above and its end date is higher than start date of period being created.
But if you go and try to create period 08.05.-15.05. your check won't return any dates and you will create a period that is overlapping.

What would I suggest to do is to get all periods for a certain property(this will not be a big number so loading in to working memory is okay)
Then iterate through all existing ones and check if any is overlapping with one being created.
How they can overlap
A---B----B----A Start date of creating >= Existing start && End date of creating <= Existing end  (inside)
A---B----A----B Start date of creating >= Existing && Start date of creating <= Existing end (starts inside period)
B---A----B----A End date of creating >= Existing start && End of creating <= Existing end (ends inside period)
B---A----A----B Start of creating <= Existing start && End of creating >= Existing end (huggs period completely)

Please think of other cases or consult with AI tools

Reply:
Implement this on top of DB using EF
Just provide both start and end date to the repository and then do where extension to check for these cases.